[Federation] Kubefed Init should use the right RBAC API version clientset

irfanurrehman opened the pull request:
**What this PR does / why we need it**:
Implements the need as described in https://github.com/kubernetes/kubernetes/issues/41263
**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #
https://github.com/kubernetes/kubernetes/issues/41263

**Special notes for your reviewer**:
@madhusudancs @shashidharatd @marun 
cc @kubernetes/sig-federation-bugs

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access) 
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`. 
-->
```
NONE
```

Commit:
[Federation][Kubefed] Use the discovered api version using the discovery client for RBAC

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/41830)
<!-- Reviewable:end -->

irfanurrehman commented:
@k8s-bot cvm gce e2e test this

irfanurrehman commented:
@k8s-bot gci gce e2e test this

irfanurrehman commented:
@k8s-bot cvm gke e2e test this

irfanurrehman commented:
@k8s-bot cvm gce e2e test this

ethernetdan commented:
A code freeze exception has been submitted for this PR - please let @kubernetes/release-team know of any status changes.

Commit:
[Federation][Kubefed] Add some faked methods on kubectl fake test API factory for kubefeds consumption

irfanurrehman commented:
@madhusudancs, Please take note.
Rebased this on latest.

irfanurrehman commented:
@k8s-bot gci gke e2e test this
@k8s-bot cvm gke e2e test this
@k8s-bot cvm gce e2e test this

irfanurrehman commented:
@k8s-bot gci gke e2e test this
@k8s-bot cvm gke e2e test this
@k8s-bot cvm gce e2e test this

irfanurrehman commented:
@madhusudancs, updated as discussed in https://github.com/kubernetes/kubernetes/issues/42352
PTAL!

irfanurrehman commented:
@k8s-bot cvm gce e2e test this

madhusudancs commented:
Reviewed 3 of 3 files at r2.
Review status: 3 of 4 files reviewed at latest revision, 10 unresolved discussions.

---

*[federation/pkg/kubefed/init/init.go, line 338 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41830#-KeaF8sGdrDVqO86kEQ6:-Ke_z1haIwURAubsZA6i:b-sqbt62) ([raw file](https://github.com/kubernetes/kubernetes/blob/140631fd55a97cd82f1b47a611ae130aee64e91a/federation/pkg/kubefed/init/init.go#L338)):*
> ```Go
> 	sa.Name = ""
> 	// 7. Create deployment for federation controller manager
> 	versionedClientset, err := getVersionedClientForRBACOrFail(hostFactory)
> ```

Is it possible to create this client somewhere at the top and call it `rbacVersionedClientset` or something? We use this clientset only for RBACs right?

---

*[federation/pkg/kubefed/init/init.go, line 340 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41830#-Ke_zYX8P7gjDspV2VDE:-Ke_zYX8P7gjDspV2VDF:bqg9b2x) ([raw file](https://github.com/kubernetes/kubernetes/blob/140631fd55a97cd82f1b47a611ae130aee64e91a/federation/pkg/kubefed/init/init.go#L340)):*
> ```Go
> 	versionedClientset, err := getVersionedClientForRBACOrFail(hostFactory)
> 	if err != nil {
> 		if err.Error() != rbacAPINotAvailable {
> ```

Please use an error type instead of a string. More concretely define:

```golang
// noRBACAPIError is a trivial implementation of error.
type noRBACAPIError struct {
    s string
}

func (n *noRBACAPIError) Error() string {
    return n.s
}
```

And in this logic here:

```golang
versionedClientset, err := getVersionedClientForRBACOrFail(hostFactory)
if rbacErr, ok := err.(*noRBACAPIError); ok {
} else if err != nil {
...
} else {
...
}
```

This is a good reference - https://blog.golang.org/error-handling-and-go

---

*[federation/pkg/kubefed/init/init.go, line 346 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41830#-KeaGiwqImsU8JR8OAUS:-KeaGiwqImsU8JR8OAUT:bsa6mk8) ([raw file](https://github.com/kubernetes/kubernetes/blob/140631fd55a97cd82f1b47a611ae130aee64e91a/federation/pkg/kubefed/init/init.go#L346)):*
> ```Go
> 		// 7a. Create a service account in the host cluster for federation
> 		// controller manager.
> 		sa, err = createControllerManagerSA(versionedClientset, i.commonOptions.FederationSystemNamespace, i.options.dryRun)
> ```

Should we create a separate service account anyway, even when RBAC support doesn't exist but just not create the roles and bindings? I don't know what the right answer is, I am just thinking about the pros and cons. What do you think?

---

*[federation/pkg/kubefed/init/init.go, line 1115 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41830#-KeaHoEaaKQZBDHZ00Lh:-KeaHoEaaKQZBDHZ00Li:bg7yigj) ([raw file](https://github.com/kubernetes/kubernetes/blob/140631fd55a97cd82f1b47a611ae130aee64e91a/federation/pkg/kubefed/init/init.go#L1115)):*
> ```Go
> }
> 
> func getVersionedClientForRBACOrFail(hostFactory cmdutil.Factory) (client.Interface, error) {
> ```

Can you move this to `kubefed/util` package? We want to reuse this in `kubefed join` too later right?

---

*[federation/pkg/kubefed/init/init.go, line 1122 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41830#-KeaI6AKG8UzOVTIFuXU:-KeaI6AKG8UzOVTIFuXV:b-xelw9r) ([raw file](https://github.com/kubernetes/kubernetes/blob/140631fd55a97cd82f1b47a611ae130aee64e91a/federation/pkg/kubefed/init/init.go#L1122)):*
> ```Go
> 	groupList, err := discoveryclient.ServerGroups()
> 	if err != nil {
> 		return nil, fmt.Errorf("Couldn't get available api versions from host cluster: %v", err)
> ```

s/Couldn't/couldn't/

---

*[federation/pkg/kubefed/init/init.go, line 1125 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41830#-KeaSN84a1uCOQ_06B8U:-KeaSN84a1uCOQ_06B8V:b-apepy9) ([raw file](https://github.com/kubernetes/kubernetes/blob/140631fd55a97cd82f1b47a611ae130aee64e91a/federation/pkg/kubefed/init/init.go#L1125)):*
> ```Go
> 	}
> 
> 	for _, g := range groupList.Groups {
> ```

Isn't this already made available by the `RESTMapper` interface? See https://github.com/kubernetes/kubernetes/blob/master/federation/pkg/kubefed/unjoin.go#L137 for an example. You can just skip specifying a version

@kubernetes/sig-api-machinery-misc isn't this the case? Shouldn't we able to use the discovered/available versioned client by just using the `RESTMapper` interface with the concrete type initialized here - https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/util/factory_object_mapping.go#L68

---

*[federation/pkg/kubefed/init/init.go, line 1152 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41830#-KeaK-11oLv9-tHVe1rN:-KeaK-11oLv9-tHVe1rO:bkohhgr) ([raw file](https://github.com/kubernetes/kubernetes/blob/140631fd55a97cd82f1b47a611ae130aee64e91a/federation/pkg/kubefed/init/init.go#L1152)):*
> ```Go
> 	}
> 
> 	return nil, fmt.Errorf(rbacAPINotAvailable)
> ```

Return the error type described above in a comment.

---

*[federation/pkg/kubefed/init/init_test.go, line 219 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41830#-KeaTjc4SePt3P4ZWQfs:-KeaTjc5He1nskOs9EPs:bin63ef) ([raw file](https://github.com/kubernetes/kubernetes/blob/140631fd55a97cd82f1b47a611ae130aee64e91a/federation/pkg/kubefed/init/init_test.go#L219)):*
> ```Go
> 		// Check pkg/kubectl/cmd/testing/fake (fakeAPIFactory.DiscoveryClient()) for details of tmpDir
> 		// We want an unique discovery cache path for each test run, else the case from previous case would be used
> 		tmpDirPath, _ = ioutil.TempDir("", "")
> ```

Error cannot be ignored here.

---

*[federation/pkg/kubefed/init/init_test.go, line 1049 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41830#-KeaVJ4W6-qG_i0oVZof:-KeaVJ4W6-qG_i0oVZog:b-33zucq) ([raw file](https://github.com/kubernetes/kubernetes/blob/140631fd55a97cd82f1b47a611ae130aee64e91a/federation/pkg/kubefed/init/init_test.go#L1049)):*
> ```Go
> 			},
> 		},
> 	}
> ```

Is ABAC actually used? Or is it just to have one more API group in the list? If it is the latter, can we use some other API type that's unrelated to access controls? It is kind of confusing now.

---

*[pkg/kubectl/cmd/testing/fake.go, line 573 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41830#-KeaX27DhkO9XfQtavRB:-KeaX27EC67n7NBBXj5-:b-w5k6s1) ([raw file](https://github.com/kubernetes/kubernetes/blob/140631fd55a97cd82f1b47a611ae130aee64e91a/pkg/kubectl/cmd/testing/fake.go#L573)):*
> ```Go
> }
> 
> func (f *fakeAPIFactory) DiscoveryClient() (discovery.CachedDiscoveryInterface, error) {
> ```

FakeFactory already has a `DiscoveryClient()` isn't that sufficient? Also, if you switch to `RESTMapper` interface, this wouldn't be necessary right?

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41830)*
<!-- Sent from Reviewable.io -->

madhusudancs commented:
Reviewed 1 of 1 files at r1.
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41830)*
<!-- Sent from Reviewable.io -->

irfanurrehman commented:
Review status: 1 of 6 files reviewed at latest revision, 10 unresolved discussions.

---

*[federation/pkg/kubefed/init/init.go, line 338 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41830#-KeaF8sGdrDVqO86kEQ6:-KeiE8Acu6yamsFWt5-J:b-nb70bu) ([raw file](https://github.com/kubernetes/kubernetes/blob/140631fd55a97cd82f1b47a611ae130aee64e91a/federation/pkg/kubefed/init/init.go#L338)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

Is it possible to create this client somewhere at the top and call it `rbacVersionedClientset` or something? We use this clientset only for RBACs right?
</blockquote></details>

Yes, this is used only for rbac. Updated the name.
Did not understand the logic of creating it somewhere at the top?

---

*[federation/pkg/kubefed/init/init.go, line 340 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41830#-Ke_zYX8P7gjDspV2VDE:-KeiF8kAS7BZwFegsFgQ:b-ho9jnp) ([raw file](https://github.com/kubernetes/kubernetes/blob/140631fd55a97cd82f1b47a611ae130aee64e91a/federation/pkg/kubefed/init/init.go#L340)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

Please use an error type instead of a string. More concretely define:

```golang
// noRBACAPIError is a trivial implementation of error.
type noRBACAPIError struct {
    s string
}

func (n *noRBACAPIError) Error() string {
    return n.s
}
```

And in this logic here:

```golang
versionedClientset, err := getVersionedClientForRBACOrFail(hostFactory)
if rbacErr, ok := err.(*noRBACAPIError); ok {
} else if err != nil {
...
} else {
...
}
```

This is a good reference - https://blog.golang.org/error-handling-and-go

</blockquote></details>

Thanks for the suggestion. I actually did consider this, but the current logic sort of worked anyways, and for code brevity.
Updating!

---

*[federation/pkg/kubefed/init/init.go, line 346 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41830#-KeaGiwqImsU8JR8OAUS:-KeiLo3Rgi4XdsJMw2w7:b-s41mo7) ([raw file](https://github.com/kubernetes/kubernetes/blob/140631fd55a97cd82f1b47a611ae130aee64e91a/federation/pkg/kubefed/init/init.go#L346)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

Should we create a separate service account anyway, even when RBAC support doesn't exist but just not create the roles and bindings? I don't know what the right answer is, I am just thinking about the pros and cons. What do you think?
</blockquote></details>

We can do that. On the other hand, I am not sure, if there is some version where the service account as a feature itself was not there. (but then the same might be true for secrets, or deployments). I think we can keep it the way it is implemented now, service account created only when mandated by the cluster for our logic to work. Thoughts?

---

*[federation/pkg/kubefed/init/init.go, line 1115 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41830#-KeaHoEaaKQZBDHZ00Lh:-KeiNtONDL0qmA5_0RKA:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/140631fd55a97cd82f1b47a611ae130aee64e91a/federation/pkg/kubefed/init/init.go#L1115)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

Can you move this to `kubefed/util` package? We want to reuse this in `kubefed join` too later right?
</blockquote></details>

Done.

---

*[federation/pkg/kubefed/init/init.go, line 1122 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41830#-KeaI6AKG8UzOVTIFuXU:-KeiO5kaD4BBrZtSRvVQ:b-2gewmi) ([raw file](https://github.com/kubernetes/kubernetes/blob/140631fd55a97cd82f1b47a611ae130aee64e91a/federation/pkg/kubefed/init/init.go#L1122)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

s/Couldn't/couldn't/
</blockquote></details>

I kept it as caps, because in case of this error, the same is returned from the command (kubefed init) directly to the user.

---

*[federation/pkg/kubefed/init/init.go, line 1125 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41830#-KeaSN84a1uCOQ_06B8U:-KeiWP6pvT2LIt_BAOgO:b-kz9u9r) ([raw file](https://github.com/kubernetes/kubernetes/blob/140631fd55a97cd82f1b47a611ae130aee64e91a/federation/pkg/kubefed/init/init.go#L1125)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

Isn't this already made available by the `RESTMapper` interface? See https://github.com/kubernetes/kubernetes/blob/master/federation/pkg/kubefed/unjoin.go#L137 for an example. You can just skip specifying a version

@kubernetes/sig-api-machinery-misc isn't this the case? Shouldn't we able to use the discovered/available versioned client by just using the `RESTMapper` interface with the concrete type initialized here - https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/util/factory_object_mapping.go#L68
</blockquote></details>

It is correct that the available group version and kinds can be found out for a concrete object (rbac in our case) using the typer and mapper interfaces. But exploring further, into the mechanism, I could not get a way to arrive at the versioned client for the preferred version.
Suggestions?
I would certainly not mind updating the implementation, if there is a better way of achieving what we are doing here.

---

*[federation/pkg/kubefed/init/init.go, line 1152 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41830#-KeaK-11oLv9-tHVe1rN:-KeiY-0fQMinmWu_XHYR:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/140631fd55a97cd82f1b47a611ae130aee64e91a/federation/pkg/kubefed/init/init.go#L1152)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

Return the error type described above in a comment.
</blockquote></details>

Done.

---

*[federation/pkg/kubefed/init/init_test.go, line 219 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41830#-KeaTjc4SePt3P4ZWQfs:-KeiZkAAim6eG5rOhiHZ:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/140631fd55a97cd82f1b47a611ae130aee64e91a/federation/pkg/kubefed/init/init_test.go#L219)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

Error cannot be ignored here.
</blockquote></details>

Done.

---

*[federation/pkg/kubefed/init/init_test.go, line 1049 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41830#-KeaVJ4W6-qG_i0oVZof:-KeiZm30W_npCouCioYW:b-evlu1m) ([raw file](https://github.com/kubernetes/kubernetes/blob/140631fd55a97cd82f1b47a611ae130aee64e91a/federation/pkg/kubefed/init/init_test.go#L1049)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

Is ABAC actually used? Or is it just to have one more API group in the list? If it is the latter, can we use some other API type that's unrelated to access controls? It is kind of confusing now.
</blockquote></details>

Yes, used it just to add another api group, using an arbitrary test API name now.

---

*[pkg/kubectl/cmd/testing/fake.go, line 573 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41830#-KeaX27DhkO9XfQtavRB:-Keia7SELIeiDn2qUij5:b-7yu2nv) ([raw file](https://github.com/kubernetes/kubernetes/blob/140631fd55a97cd82f1b47a611ae130aee64e91a/pkg/kubectl/cmd/testing/fake.go#L573)):*
> FakeFactory 
to me it appeared like FakeFactory and fakeAPIFactory are two different structures (probably both faking cmdutil.Factory); am I not able to understand something?
Also, please check the comment about switching to RESTMapper.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41830)*
<!-- Sent from Reviewable.io -->

irfanurrehman commented:
@k8s-bot non-cri e2e test this

madhusudancs commented:
Reviewed 5 of 5 files at r3.
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---

*[federation/pkg/kubefed/init/init.go, line 338 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41830#-KeaF8sGdrDVqO86kEQ6:-KejgSBX2n9hgicc80hf:bo8vwyq) ([raw file](https://github.com/kubernetes/kubernetes/blob/140631fd55a97cd82f1b47a611ae130aee64e91a/federation/pkg/kubefed/init/init.go#L338)):*
<details><summary><i>Previously, irfanurrehman wrote…</i></summary><blockquote>

Yes, this is used only for rbac. Updated the name.
Did not understand the logic of creating it somewhere at the top?
</blockquote></details>

I meant, initialize all the clients at the beginning of this function, but ignore it. This is fine too because this is the only place where you use this client and we generally want the initialization to be closer to its usage.

---

*[federation/pkg/kubefed/init/init.go, line 340 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41830#-Ke_zYX8P7gjDspV2VDE:-Kek3Sh-jKoD98-3yX0U:be2lqn4) ([raw file](https://github.com/kubernetes/kubernetes/blob/140631fd55a97cd82f1b47a611ae130aee64e91a/federation/pkg/kubefed/init/init.go#L340)):*
<details><summary><i>Previously, irfanurrehman wrote…</i></summary><blockquote>

Thanks for the suggestion. I actually did consider this, but the current logic sort of worked anyways, and for code brevity.
Updating!
</blockquote></details>

The previous logic was too dependent on strings. More type-safety the better.

Could you also leave a comment that says if the error is due to non-existence RBAC API, then we just continue without creating service accounts. It is not obvious right now.

---

*[federation/pkg/kubefed/init/init.go, line 346 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41830#-KeaGiwqImsU8JR8OAUS:-Kek3zS06jYtARVkVz6j:b-yjlwrg) ([raw file](https://github.com/kubernetes/kubernetes/blob/140631fd55a97cd82f1b47a611ae130aee64e91a/federation/pkg/kubefed/init/init.go#L346)):*
<details><summary><i>Previously, irfanurrehman wrote…</i></summary><blockquote>

We can do that. On the other hand, I am not sure, if there is some version where the service account as a feature itself was not there. (but then the same might be true for secrets, or deployments). I think we can keep it the way it is implemented now, service account created only when mandated by the cluster for our logic to work. Thoughts?
</blockquote></details>

A lot of things in Kubernetes won't work without service accounts. If you disable service accounts, then no application running inside a pod can talk to the API Server without using credentials. That also means many of the kubernetes components won't work.

But anyway, let's leave it this way for now. Can you add a TODO that says, "we must evaluate creating a separate service account even RBAC APIs don't exist"?

---

*[federation/pkg/kubefed/init/init.go, line 1122 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41830#-KeaI6AKG8UzOVTIFuXU:-Kek77zIxtK4c5i2lqt8:b-7zf90w) ([raw file](https://github.com/kubernetes/kubernetes/blob/140631fd55a97cd82f1b47a611ae130aee64e91a/federation/pkg/kubefed/init/init.go#L1122)):*
<details><summary><i>Previously, irfanurrehman wrote…</i></summary><blockquote>

I kept it as caps, because in case of this error, the same is returned from the command (kubefed init) directly to the user.
</blockquote></details>

This shouldn't be a user facing error this way.

What does it mean for a user to see "Couldn't get available api versions from host cluster"? Which API? What version? The user facing error should say something like ("Couldn't create RBAC roles in the host cluster: %v", err). Makes sense?

---

*[federation/pkg/kubefed/init/init.go, line 1125 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41830#-KeaSN84a1uCOQ_06B8U:-Kek6d3Bneupit5E7cM_:bwz3zaq) ([raw file](https://github.com/kubernetes/kubernetes/blob/140631fd55a97cd82f1b47a611ae130aee64e91a/federation/pkg/kubefed/init/init.go#L1125)):*
<details><summary><i>Previously, irfanurrehman wrote…</i></summary><blockquote>

It is correct that the available group version and kinds can be found out for a concrete object (rbac in our case) using the typer and mapper interfaces. But exploring further, into the mechanism, I could not get a way to arrive at the versioned client for the preferred version.
Suggestions?
I would certainly not mind updating the implementation, if there is a better way of achieving what we are doing here.
</blockquote></details>

You won't get a versioned client, but you get a rest client. You can then pass that rest client and the mapping to a resource helper `pkg/kubectl/resource.NewHelper()` and use that helper to create the required RBAC objects. Anyway, because this is a bug fix that should get into v1.6, just leave it the code this way. But please leave a TODO to evaluate the RESTMapper approach.

---

*[federation/pkg/kubefed/util/util.go, line 235 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41830#-Kek7qjeA1edTNCUF-nm:-Kek7qjeA1edTNCUF-nn:b-twwchr) ([raw file](https://github.com/kubernetes/kubernetes/blob/739bd203f69a877a93720158c73056d44ce36dce/federation/pkg/kubefed/util/util.go#L235)):*
> ```Go
> 	groupList, err := discoveryclient.ServerGroups()
> 	if err != nil {
> 		return nil, fmt.Errorf("Couldn't get available api versions from host cluster: %v", err)
> ```

s/Couldn't/couldn't/

(See comment above)

---

*[federation/pkg/kubefed/util/util.go, line 243 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41830#-Kek7nFY420wW2NMTwTB:-Kek7nFY420wW2NMTwTC:b7qeyqd) ([raw file](https://github.com/kubernetes/kubernetes/blob/739bd203f69a877a93720158c73056d44ce36dce/federation/pkg/kubefed/util/util.go#L243)):*
> ```Go
> 				gv := strings.Split(g.PreferredVersion.GroupVersion, "/")
> 				if len(gv) != 2 {
> 					return nil, fmt.Errorf("Incorrect api groupversion received from host cluster")
> ```

s/Incorrect/incorrect/

---

*[federation/pkg/kubefed/util/util.go, line 252 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41830#-Kek81HnR1Xe1d4lAZlQ:-Kek81HnR1Xe1d4lAZlR:b3nywpu) ([raw file](https://github.com/kubernetes/kubernetes/blob/739bd203f69a877a93720158c73056d44ce36dce/federation/pkg/kubefed/util/util.go#L252)):*
> ```Go
> 			for i := 0; i < len(g.Versions); i++ {
> 				if g.Versions[i].GroupVersion != "" {
> 					gv := strings.Split(g.Versions[i].GroupVersion, "/")
> ```

I would have liked to use the RESTMapper interface. But given that we are not going use that now, can you use the `ParseGroupVersion` function from https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/runtime/schema/group_version.go#L192 and use the returned object below?

---

*[federation/pkg/kubefed/util/util.go, line 254 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41830#-Kek7kIxS-s8HPkbGOD3:-Kek7kIxS-s8HPkbGOD4:b7qeyqd) ([raw file](https://github.com/kubernetes/kubernetes/blob/739bd203f69a877a93720158c73056d44ce36dce/federation/pkg/kubefed/util/util.go#L254)):*
> ```Go
> 					gv := strings.Split(g.Versions[i].GroupVersion, "/")
> 					if len(gv) != 2 {
> 						return nil, fmt.Errorf("Incorrect api groupversion received from host cluster")
> ```

s/Incorrect/incorrect/

---

*[pkg/kubectl/cmd/testing/fake.go, line 573 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41830#-KeaX27DhkO9XfQtavRB:-Kek8kOzbb5-QEQtaRUO:b8ygluc) ([raw file](https://github.com/kubernetes/kubernetes/blob/140631fd55a97cd82f1b47a611ae130aee64e91a/pkg/kubectl/cmd/testing/fake.go#L573)):*
<details><summary><i>Previously, irfanurrehman wrote…</i></summary><blockquote>

> FakeFactory 
to me it appeared like FakeFactory and fakeAPIFactory are two different structures (probably both faking cmdutil.Factory); am I not able to understand something?
Also, please check the comment about switching to RESTMapper.
</blockquote></details>

Question is what led to adding this method? Did the compilation fail without this? Or the test failed? I am trying to understand the reason behind this addition.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41830)*
<!-- Sent from Reviewable.io -->

irfanurrehman commented:
Review status: 4 of 6 files reviewed at latest revision, 10 unresolved discussions.

---

*[federation/pkg/kubefed/init/init.go, line 338 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41830#-KeaF8sGdrDVqO86kEQ6:-KelhMpuPAIVrM0pPvrB:b-t77xdl) ([raw file](https://github.com/kubernetes/kubernetes/blob/140631fd55a97cd82f1b47a611ae130aee64e91a/federation/pkg/kubefed/init/init.go#L338)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

I meant, initialize all the clients at the beginning of this function, but ignore it. This is fine too because this is the only place where you use this client and we generally want the initialization to be closer to its usage.
</blockquote></details>

agree.

---

*[federation/pkg/kubefed/init/init.go, line 340 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41830#-Ke_zYX8P7gjDspV2VDE:-Kela5eQfldZ_LzotIK8:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/140631fd55a97cd82f1b47a611ae130aee64e91a/federation/pkg/kubefed/init/init.go#L340)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

The previous logic was too dependent on strings. More type-safety the better.

Could you also leave a comment that says if the error is due to non-existence RBAC API, then we just continue without creating service accounts. It is not obvious right now.
</blockquote></details>

Done.

---

*[federation/pkg/kubefed/init/init.go, line 346 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41830#-KeaGiwqImsU8JR8OAUS:-KelbNc3wV-D4kdGNZIU:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/140631fd55a97cd82f1b47a611ae130aee64e91a/federation/pkg/kubefed/init/init.go#L346)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

A lot of things in Kubernetes won't work without service accounts. If you disable service accounts, then no application running inside a pod can talk to the API Server without using credentials. That also means many of the kubernetes components won't work.

But anyway, let's leave it this way for now. Can you add a TODO that says, "we must evaluate creating a separate service account even RBAC APIs don't exist"?
</blockquote></details>

Done.

---

*[federation/pkg/kubefed/init/init.go, line 1122 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41830#-KeaI6AKG8UzOVTIFuXU:-KelbVHmaPTjsKaQ0N8V:b-q4ht8) ([raw file](https://github.com/kubernetes/kubernetes/blob/140631fd55a97cd82f1b47a611ae130aee64e91a/federation/pkg/kubefed/init/init.go#L1122)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

This shouldn't be a user facing error this way.

What does it mean for a user to see "Couldn't get available api versions from host cluster"? Which API? What version? The user facing error should say something like ("Couldn't create RBAC roles in the host cluster: %v", err). Makes sense?
</blockquote></details>

Yes it does, I have updated the error message.
This also puts another thought/question.
We have reached here because our unversioned clientset (hostClientset) worked so far, and created rest of the object until now.
Does it makes sense to continue creating rest of the objects which can be created using the hostClientset on any error here (rather then exit, that way we exit now).. ?
(Its highly unlikely, however that we might stumble on this or any other error in getting the versioned client, except that the rbac api is not available)

---

*[federation/pkg/kubefed/init/init.go, line 1125 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41830#-KeaSN84a1uCOQ_06B8U:-KeldNbjyGM38y2ptd7D:b-h6j61m) ([raw file](https://github.com/kubernetes/kubernetes/blob/140631fd55a97cd82f1b47a611ae130aee64e91a/federation/pkg/kubefed/init/init.go#L1125)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

You won't get a versioned client, but you get a rest client. You can then pass that rest client and the mapping to a resource helper `pkg/kubectl/resource.NewHelper()` and use that helper to create the required RBAC objects. Anyway, because this is a bug fix that should get into v1.6, just leave it the code this way. But please leave a TODO to evaluate the RESTMapper approach.
</blockquote></details>

Got it, thanks for the explaination!
updated the TODO

---

*[federation/pkg/kubefed/util/util.go, line 235 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41830#-Kek7qjeA1edTNCUF-nm:-Kele62P-sryO4Vm-rvf:b-bu91xm) ([raw file](https://github.com/kubernetes/kubernetes/blob/739bd203f69a877a93720158c73056d44ce36dce/federation/pkg/kubefed/util/util.go#L235)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

s/Couldn't/couldn't/

(See comment above)
</blockquote></details>

Updated

---

*[federation/pkg/kubefed/util/util.go, line 243 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41830#-Kek7nFY420wW2NMTwTB:-KeleJHVDc1bFOqIYN_c:b-uf125s) ([raw file](https://github.com/kubernetes/kubernetes/blob/739bd203f69a877a93720158c73056d44ce36dce/federation/pkg/kubefed/util/util.go#L243)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

s/Incorrect/incorrect/
</blockquote></details>

this error string is no longer needed because we use schema.ParseGroupVersion() now

---

*[federation/pkg/kubefed/util/util.go, line 252 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41830#-Kek81HnR1Xe1d4lAZlQ:-Kelh3S7_IERubDOzKsl:b-biws3u) ([raw file](https://github.com/kubernetes/kubernetes/blob/739bd203f69a877a93720158c73056d44ce36dce/federation/pkg/kubefed/util/util.go#L252)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

I would have liked to use the RESTMapper interface. But given that we are not going use that now, can you use the `ParseGroupVersion` function from https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/runtime/schema/group_version.go#L192 and use the returned object below?
</blockquote></details>

Thanks for the suggestion, updated!

---

*[federation/pkg/kubefed/util/util.go, line 254 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41830#-Kek7kIxS-s8HPkbGOD3:-Kelh6KqQn4K6S_SnZdC:b-uf125s) ([raw file](https://github.com/kubernetes/kubernetes/blob/739bd203f69a877a93720158c73056d44ce36dce/federation/pkg/kubefed/util/util.go#L254)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

s/Incorrect/incorrect/
</blockquote></details>

this error string is no longer needed because we use schema.ParseGroupVersion() now

---

*[pkg/kubectl/cmd/testing/fake.go, line 573 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41830#-KeaX27DhkO9XfQtavRB:-KelZIqt5cWySnL4Skal:b-5kgaah) ([raw file](https://github.com/kubernetes/kubernetes/blob/140631fd55a97cd82f1b47a611ae130aee64e91a/pkg/kubectl/cmd/testing/fake.go#L573)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

Question is what led to adding this method? Did the compilation fail without this? Or the test failed? I am trying to understand the reason behind this addition.
</blockquote></details>

the code uses the discovery client, but it was not faked in fakeAPIFactory, so the original cmdutil.Factory's DiscoveryClient() is invoked in the unit test   (fakeAPIFactory extends cmdutil.Factory), which does not work in unit tests.
fakeAPIFactory overrides only few methods (but not all), so I added one more override, needed by us.
I am not completely sure, but I think implementation of fakeAPIFactory was targeted to override only those methods, which probably were needed in the kubectl unit testing (or for only those which are unit tested).

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41830)*
<!-- Sent from Reviewable.io -->

madhusudancs commented:
A couple of minor nits. Let me know once you fix them, I will add the labels.

---

Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[federation/pkg/kubefed/init/init.go, line 1122 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41830#-KeaI6AKG8UzOVTIFuXU:-KemAnSObpqtuN0BzoWI:b-6tupua) ([raw file](https://github.com/kubernetes/kubernetes/blob/140631fd55a97cd82f1b47a611ae130aee64e91a/federation/pkg/kubefed/init/init.go#L1122)):*
<details><summary><i>Previously, irfanurrehman wrote…</i></summary><blockquote>

Yes it does, I have updated the error message.
This also puts another thought/question.
We have reached here because our unversioned clientset (hostClientset) worked so far, and created rest of the object until now.
Does it makes sense to continue creating rest of the objects which can be created using the hostClientset on any error here (rather then exit, that way we exit now).. ?
(Its highly unlikely, however that we might stumble on this or any other error in getting the versioned client, except that the rbac api is not available)
</blockquote></details>

Good point. This was actually the reason why I wanted the versioned clientset to be initialized at the top, so that if the initialization failed we could just give up instead of leaving things in a bad state. I think if you actually check this before creating any object, it should be still Ok to error and exit.

---

*[federation/pkg/kubefed/util/util.go, line 235 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41830#-Kek7qjeA1edTNCUF-nm:-KemBVsBSP4AZW0dn7Oe:b-jn7fl0) ([raw file](https://github.com/kubernetes/kubernetes/blob/739bd203f69a877a93720158c73056d44ce36dce/federation/pkg/kubefed/util/util.go#L235)):*
<details><summary><i>Previously, irfanurrehman wrote…</i></summary><blockquote>

Updated
</blockquote></details>

I am still seeing "Couldn't", the capitalized version. Is the update pushed?

---

*[pkg/kubectl/cmd/testing/fake.go, line 573 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41830#-KeaX27DhkO9XfQtavRB:-KemC5WAMelWDiYe-yVx:b-pu4g1r) ([raw file](https://github.com/kubernetes/kubernetes/blob/140631fd55a97cd82f1b47a611ae130aee64e91a/pkg/kubectl/cmd/testing/fake.go#L573)):*
<details><summary><i>Previously, irfanurrehman wrote…</i></summary><blockquote>

the code uses the discovery client, but it was not faked in fakeAPIFactory, so the original cmdutil.Factory's DiscoveryClient() is invoked in the unit test   (fakeAPIFactory extends cmdutil.Factory), which does not work in unit tests.
fakeAPIFactory overrides only few methods (but not all), so I added one more override, needed by us.
I am not completely sure, but I think implementation of fakeAPIFactory was targeted to override only those methods, which probably were needed in the kubectl unit testing (or for only those which are unit tested).
</blockquote></details>

Ok fair enough.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41830#-:-KemCHvtql6C6a_O3vRH:b-ehzhap)*
<!-- Sent from Reviewable.io -->

irfanurrehman commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[federation/pkg/kubefed/init/init.go, line 1122 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41830#-KeaI6AKG8UzOVTIFuXU:-KemLbi-1v4qYal2Q-av:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/140631fd55a97cd82f1b47a611ae130aee64e91a/federation/pkg/kubefed/init/init.go#L1122)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

Good point. This was actually the reason why I wanted the versioned clientset to be initialized at the top, so that if the initialization failed we could just give up instead of leaving things in a bad state. I think if you actually check this before creating any object, it should be still Ok to error and exit.
</blockquote></details>

Done.

---

*[federation/pkg/kubefed/util/util.go, line 235 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41830#-Kek7qjeA1edTNCUF-nm:-KemNSP6l71kNKOdMj2X:b-h51i11) ([raw file](https://github.com/kubernetes/kubernetes/blob/739bd203f69a877a93720158c73056d44ce36dce/federation/pkg/kubefed/util/util.go#L235)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

I am still seeing "Couldn't", the capitalized version. Is the update pushed?
</blockquote></details>

Its because I did update the error message to the user facing error message as per your suggestion.. :-)
"Couldn't create RBAC roles in the host cluster: %v", err

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41830)*
<!-- Sent from Reviewable.io -->

irfanurrehman commented:
Review status: 5 of 6 files reviewed at latest revision, 2 unresolved discussions.

---

*[federation/pkg/kubefed/util/util.go, line 235 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41830#-Kek7qjeA1edTNCUF-nm:-KemOoOnpk_jIOeCngZX:bpzlnjx) ([raw file](https://github.com/kubernetes/kubernetes/blob/739bd203f69a877a93720158c73056d44ce36dce/federation/pkg/kubefed/util/util.go#L235)):*
<details><summary><i>Previously, irfanurrehman wrote…</i></summary><blockquote>

Its because I did update the error message to the user facing error message as per your suggestion.. :-)
"Couldn't create RBAC roles in the host cluster: %v", err
</blockquote></details>

on second thoughts should I update this to -> 
"Couldn't get clientset to create RBAC roles in the host cluster"
because if this happens, we will fail without doing anything..

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41830)*
<!-- Sent from Reviewable.io -->

liggitt commented:
just because the RBAC API is available doesn't mean the RBAC authorizer is in use by the underlying cluster. `kubefed` must always have an option to be given credentials to use for the underlying cluster, and use them even when the RBAC API is present

irfanurrehman commented:
Review status: 5 of 6 files reviewed at latest revision, 2 unresolved discussions.

---

*[federation/pkg/kubefed/util/util.go, line 235 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41830#-Kek7qjeA1edTNCUF-nm:-KencOFkAjU9eXz6GGfC:bsxdby9) ([raw file](https://github.com/kubernetes/kubernetes/blob/739bd203f69a877a93720158c73056d44ce36dce/federation/pkg/kubefed/util/util.go#L235)):*
<details><summary><i>Previously, irfanurrehman wrote…</i></summary><blockquote>

on second thoughts should I update this to -> 
"Couldn't get clientset to create RBAC roles in the host cluster"
because if this happens, we will fail without doing anything..
</blockquote></details>

I have gone ahead and changed the error message, please let me know if u have a different suggestion

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41830)*
<!-- Sent from Reviewable.io -->

marun commented:
@madhusudancs @irfanurrehman Why was this not targeted at 1.6, with an exception if necessary?  I think this qualifies as a bug/regression given that kubefed will no longer be able to deploy to clusters without rbac support (< 1.6).

marun commented:
Review status: 4 of 6 files reviewed at latest revision, 3 unresolved discussions.

---

*[federation/pkg/kubefed/init/init.go, line 264 at r5](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41830#-KenkPNQ12yMPZp2gCfh:-KenkPNQ12yMPZp2gCfi:bya0ekr) ([raw file](https://github.com/kubernetes/kubernetes/blob/a2d6d98c68b017831ab914613412eccd5792f177/federation/pkg/kubefed/init/init.go#L264)):*
> ```Go
> 	}
> 
> 	rbacAvailable := true
> ```

Consider moving the rbac-related code into its own function that can be tested independently.  This function is too long as it is.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41830)*
<!-- Sent from Reviewable.io -->

irfanurrehman commented:
@k8s-bot kops aws e2e test this

madhusudancs commented:
> just because the RBAC API is available doesn't mean the RBAC authorizer is in use by the underlying cluster. kubefed must always have an option to be given credentials to use for the underlying cluster, and use them even when the RBAC API is present

@liggitt not entirely sure I completely understand this. Could you please elaborate? 

We did not think about authorizer's availability. Where can we read more about it? By credentials, do you mean user credentials? If we are to give user credentials, what's the point of service accounts then?

madhusudancs commented:
@liggitt we would still like to discuss your suggestion and implement the required changes. But I think those changes should come in a follow up PR.

I will let this PR merge so people can try this out on 1.5 or older clusters.

---

Reviewed 1 of 2 files at r5.
Review status: 5 of 6 files reviewed at latest revision, 1 unresolved discussion.

---

*[federation/pkg/kubefed/init/init.go, line 264 at r5](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41830#-KenkPNQ12yMPZp2gCfh:-KerDn_Zh9GEUz6zx6cP:b6313d4) ([raw file](https://github.com/kubernetes/kubernetes/blob/a2d6d98c68b017831ab914613412eccd5792f177/federation/pkg/kubefed/init/init.go#L264)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

Consider moving the rbac-related code into its own function that can be tested independently.  This function is too long as it is.
</blockquote></details>

@irfanurrehman this comment makes sense. I am applying the LGTM label now.  Please send another PR to refactor this. I would really like to get this in soon.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41830#-:-KerDN5hQQhhoQ0WpQCX:buyg4a2)*
<!-- Sent from Reviewable.io -->

madhusudancs commented:
/lgtm
/approve

---

Reviewed 1 of 2 files at r5.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41830#-:-KerE2mQkTy1eozlG9iC:b6yv7yi)*
<!-- Sent from Reviewable.io -->

madhusudancs commented:
@kubernetes/kubernetes-release-managers @ethernetdan @calebamiles This is a fix for a regression introduced in v1.6 cycle. We filed an exception for #42042 and this PR was a prerequisite for #42042, so we listed this PR as well. The exception request was rejected, but we still need this to fix the regression.

We will get #42042 in, in v1.7 cycle. But I am removing the `do-not-merge` label and applying `kind/bug` for this in to get this in.

madhusudancs commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[federation/pkg/kubefed/init/init.go, line 264 at r5](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41830#-KenkPNQ12yMPZp2gCfh:-KerGMSX5S5UlzK1tehl:b1jg9s1) ([raw file](https://github.com/kubernetes/kubernetes/blob/a2d6d98c68b017831ab914613412eccd5792f177/federation/pkg/kubefed/init/init.go#L264)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary><blockquote>

@irfanurrehman this comment makes sense. I am applying the LGTM label now.  Please send another PR to refactor this. I would really like to get this in soon.
</blockquote></details>

I just realized that this PR also needs an approval from someone in `pkg/kubectl/OWNERS`. That might not happen until tomorrow morning PST. If you have time and make this change here, please let me know. I will apply the lgtm label again.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/41830)*
<!-- Sent from Reviewable.io -->

irfanurrehman commented:
@marun, @madhusudancs 
I am leaving the implementation as is now.
1 - we need 2 distinct functionality in the same function
  a - to try and get the versioned client at the beginning of init to fail fast.
  b - create the SA and roles in sequence at the same place as we do now. 
2 - There already are TODOs which need to be addressed especially for RBAC related functionality.

Will take this as a follow up.

liggitt commented:
> @liggitt not entirely sure I completely understand this. Could you please elaborate?
> 
> We did not think about authorizer's availability. Where can we read more about it? By credentials, do you mean user credentials? If we are to give user credentials, what's the point of service accounts then?

`kubefed` does not know what authorizer the underlying cluster is using (it could be using ABAC, RBAC, or a webhook authorizer). It is also not a given that the cluster operator wants a service account to have cluster-admin-level powers. In both of those cases, the cluster operator would be responsible for giving the federation server a credential to use, and granting the appropriate permissions to the user that credential identifies.

`kubefed` having the ability to provision a service account and set up RBAC rules is nice, and will likely be used in many cases, but should not be the only allowed method of joining.

I see the authn/authz step as a pre-step to joining, one which can either be satisfied by the operator explicitly providing kubefed a credential to use (and being responsible for having granted the appropriate permissions to it out of band), or by kubefed setting up a service account and RBAC grant.

liggitt commented:
maybe I'm misunderstanding which server this is being created against... is this creating service accounts and rbac grants in the server hosting the federation deployment, or in the server being joined to the federated cluster?

madhusudancs commented:
> maybe I'm misunderstanding which server this is being created against... is this creating service accounts and rbac grants in the server hosting the federation deployment, or in the server being joined to the federated cluster?

This PR as it stands is targeting the cluster that hosts the federation control plane. 

Service accounts for joining clusters is not going in until 1.7. So this only affects host clusters.

What you describe makes sense. However, it's scope is larger than a bug fix. I will file an issue for this and hope to get this resolved in v1.7 along with joining clusters service accounts. I will let this PR in for now though.

madhusudancs commented:
> Will take this as a follow up.

Follow up is fine. Just a minor note, you don't have to put the service account creation part in a separate function. If you just move the clientset creation part to a new function that returns the client and the bool, that should be fine. You are already doing effectively that with the other function in `utils` that returns the clientset and an error. This just another layer of abstraction to interpret what the error means.

pwittrock commented:
/approve
